### PR TITLE
FIX: Extra parameters, custom proj

### PR DESCRIPTION
**Pull request by Eric89GXL**
Adds an entry point for custom projectors, along with a couple other minor options.

### CODE REVIEW

**Comment by Eric89GXL** [comment]
@kambysese can you have a look and see if it makes sense to you?

**Comment by Eric89GXL** [comment]
@rkmaddox feel free to have a look as well

File: mnefun/_mnefun.py [hunk 1]
```diff
@@ -152,6 +152,14 @@ class Params(object):
         be used for covariance calculation. This is useful e.g. when using
         a high-pass filter and no baselining is desired (but evoked
         covariances should still be calculated from the baseline period).
+    reject_tmin : float | None
+        Reject minimum time to use when epoching. None will use ``tmin``.
+    reject_tmax : float | None
+        Reject maximum time to use when epoching. None will use ``tmax``.
```
Comment (kambysese) [inline mnefun/_mnefun.py]: This is presumably for rejection restricted to an interval WITHIN the epoching interval defined by `tmin` - `tmax`?

File: mnefun/_mnefun.py [hunk 2]
```diff
@@ -1400,8 +1417,7 @@ def do_preprocessing_combined(p, subjects, run_indices):
         bad_dir = op.join(p.work_dir, subj, p.bad_dir)
 
         # Create SSP projection vectors after marking bad channels
-        raw_names = get_raw_fnames(p, subj, 'sss', False, False,
-                                   run_indices[si])
+        raw_names = get_raw_fnames(p, subj, 'sss', False)
```
Comment (kambysese) [inline mnefun/_mnefun.py]: Why get rid of `run_indices`?

File: mnefun/_mnefun.py [hunk 3]
```diff
@@ -1492,8 +1514,17 @@ def do_preprocessing_combined(p, subjects, run_indices):
...
-                              filter_length=p.filter_length, force_bads=False)
+                              filter_length=p.filter_length, force_bads=False,
+                              l_trans=p.hp_trans, h_trans=p.lp_trans)
+
+        # Apply any user-supplied extra projectors
+        if p.proj_extra is not None:
+            if p.disp_files:
+                print('    Adding extra projectors from "%s".' % p.proj_extra)
+            extra_proj = op.join(pca_dir, p.proj_extra)
+            projs = read_proj(extra_proj)
```
Comment (kambysese) [inline mnefun/_mnefun.py]: :+1: out of curiosity what would be a e.g., custom/user proj?

File: mnefun/_mnefun.py [hunk 4]
```diff
@@ -1678,22 +1715,7 @@ class FakeEpochs():
...
-        self._current = 0
-        return self
-
-    def next(self):
-        if self._current >= len(self._data):
-            raise StopIteration
-        epoch = self._data[self._current]
-        self._current += 1
-        return epoch
+        raise RuntimeError('Use mne.EpochsArray instead')
```
Comment (kambysese) [inline mnefun/_mnefun.py]: I don't understand this. It's a class function that does what beside `raise RuntimeError`?

File: mnefun/_mnefun.py [hunk 5]
```diff
@@ -1875,7 +1897,10 @@ def plot_raw_psd(p, subjects, run_indices=None, tmin=0., fmin=2, n_fft=2048):
                 warnings.warn('Unable to find %s data file.' % file_type)
             with warnings.catch_warnings(record=True):
                 raw = Raw(fname, preload=True, allow_maxshield=True)
-            fmax = p.lp_cut if file_type == 'pca' else raw.info['lowpass'] + 50
+            if file_type == 'pca':
+                fmax = p.lp_cut
+            else:
+                fmax = raw.info['lowpass'] + 50
```
Comment (kambysese) [inline mnefun/_mnefun.py]: So is the block code syntax more pythonic than the one-liner?

**Comment by kambysese** [comment]
Overall it all makes sense. BTW are the filter transition bands additions reflecting changes in mne-python? How was filtering being done before this PR?

File: mnefun/_mnefun.py (hunk 2 above)
Comment (Eric89GXL) [inline mnefun/_mnefun.py]: whoops!

File: mnefun/_mnefun.py (hunk 3 above)
Comment (Eric89GXL) [inline mnefun/_mnefun.py]: We have a paradigm where users respond verbally at the end of each trial. We added a projector to get rid of the large artifact, as it bled into the baseline period a bit due to the acquisition high-pass filter.

File: mnefun/_mnefun.py (hunk 4 above)
Comment (Eric89GXL) [inline mnefun/_mnefun.py]: Nothing. It's a really old class that I used a couple times a number of years ago. This is just a reminder to update code if I try to run it again.

File: mnefun/_mnefun.py (hunk 5 above)
Comment (Eric89GXL) [inline mnefun/_mnefun.py]: No, the line used to have some parens that made it > 80 char, and this was the other fix I had done. Doesn't really matter much either way.

File: mnefun/_mnefun.py (hunk 1 above)
Comment (Eric89GXL) [inline mnefun/_mnefun.py]: yep

**Comment by Eric89GXL** [comment]
The filtering by default has a trans_bandwidth of 0.5 Hz, so this preserves the old behavior if the numbers aren't changed.

Comments addressed.

**Comment by Eric89GXL** [comment]
@rkmaddox can you have a look, too? Might be good to have a couple of sets of eyes on this one.

**Comment by Eric89GXL** [comment]
@rkmaddox do you have time to take a look? This contains a number of fixes now.

**Comment by rkmaddox** [comment]
Yes, but not today. Gchat me to tomorrow to remind me?

**Comment by rkmaddox** [comment]
Testing failed locally because: `ImportError: No module named mock`

Do we need to add mock as a dependency?

**Comment by Eric89GXL** [comment]
I wouldn't worry too much about the tests, they are very minimal at this point

File: mnefun/_mnefun.py [hunk 6]
```diff
@@ -1642,58 +1724,19 @@ def apply_preprocessing_combined(p, subjects, run_indices):
...
-                             filter_length=p.filter_length, force_bads=False)
+                             filter_length=p.filter_length, force_bads=False,
+                             l_trans=p.hp_trans, h_trans=p.lp_trans)
             raw.save(o, overwrite=True)
         # look at raw_clean for ExG events
         if p.plot_raw:
             _viz_raw_ssp_events(p, subj, run_indices[si])
 
 
-def gen_layouts(p, subjects):
```
Comment (rkmaddox) [inline mnefun/_mnefun.py]: Has this been obsoleted by something else?

File: mnefun/_status.py [hunk 7]
```diff
@@ -102,11 +105,11 @@ def print_proc_status(p, subjects, structurals, analyses, run_indices):
             these_steps.append('missing')
         steps_all.append(these_steps)
 
-    steps = ['raw', 'sco', 'pbd', 'coreg',
```
Comment (rkmaddox) [inline mnefun/_status.py]: These are a bit obfuscated: `asp` is pretty far from "apply ssp".

**Comment by rkmaddox** [comment]
Other than the comment, LGTM. But I must admit that I find github's git display almost impossible to read...

File: mnefun/_mnefun.py (hunk 6 above)
Comment (Eric89GXL) [inline mnefun/_mnefun.py]: Yeah, mne-python has something that does this. I also doubt it was used by anyone other than me.

File: mnefun/_status.py (hunk 7 above)
Comment (Eric89GXL) [inline mnefun/_status.py]: yeah I know, but the column widths are determined by the shortest column, so 3 char is the limit

**Comment by rkmaddox** [comment]
OK. +1 for merge.

**Comment by Eric89GXL** [comment]
Thanks @rkmaddox @kambysese for the reviews

**Pull request merged**